src/garage/tf/policies/categorical_mlp_policy.py

```diff
@@ -12,7 +12,7 @@
 from garage.tf.policies.policy import StochasticPolicy
 
 
-class CategoricalMLPPolicy(StochasticPolicy):
+class CategoricalMLPPolicy(CategoricalMLPModel, StochasticPolicy):
```

Comment:
what purpose does inheriting from StochasticPolicy serve here?

Reply:
Complying with the Policy interface and indicating this policy has a `distribution` (from `StochasticPolicy`)

Reply:
Maybe instead we should move this up into StochasticModel? Then `class CategoricalMLPModel(StochasticModel)` and `class CategoricalMLPPolicy(CategorialMLPModel)`?

I think this is close to what @krzentner is trying to do in the torch tree, so you *only* have to implement the model to make a policy less boilerplate)

Reply:
Totally agree, will fix

Reply:
`CategoricalMLPModel` now inherits from `MLPModel`. In tf tree, it feels like appending `Stochastic` to a pure interface (which is `tf.Policy`) is simpler?

Reply:
Otherwise, we will have something like `CategoricalMLPModel(MLPModel, Stochastic)`

Reply:
or we can have another abc “StochasticMLPModel”

Reply:
After a second thought, I think we could get rid of `StochasticPolicy` and cache the distribution returned when calling `model.build()`. Eventually, a model does not own a distribution object -- it returns one when you call `.build()`/`.forward()`

Reply:
it will probably take multiple passes to make this work really well. for now, is suggest going with whatever gets you a simple linear inheritance pattern. i think the one i propose above (StochasticModel/CategoricalMLPModel) leads to the least number of classes in the interim. if an algorithm wants to know if something is stochastic, it can still ask `isinstance(policy, StochasticModel)` which is IMO a step forward, because the policy is now a model.

i would very much like to get rid of StochasticPolicy.

multiple inheritance (i.e. `CategoricalMLPModel(MLPModel, Stochastic)`) can get really hard to understand in python. the pattern you are describing is called the mixins pattern, but we've chosen not to use it because it makes it difficult for users to extend existing primitives, and also difficult to test.

in the torch tree, KR is working towards a future where to define a new Policy/QFunction/etc. you *only* have to define the model (i.e. `__init__` and `torch.nn.Module.forward()`, see #1528 ). I think we should try to start moving in that direction herein TF-land, too.

Reply:
In order to move towards `build-by-call` and only have to define a model for a policy, we shouldn't give policies/model default attributes, i.e. distribution. Then we can safely remove `StochasticPolicy` and when someone wants access to the distribution, they should call `.build()` and cache the network returned.

Reply:
Great point!

Reply:
So we should aim to remove stochastic entirely?

Reply:
I was thinking to abandon the use of `.distribution` at all. When someone wants a symbolic distribution, they can get it by calling `.build()` and cache the return.

Reply:
Yes, I agree! I'm just pointing out that means deterministic policies will have a different return signature from stochastic ones.